### PR TITLE
Application ExitMode

### PR DESCRIPTION
**Pull request by Gillibald**
- What does the pull request do?
This PR introduces a way to control when the application exits. You can either let the application exit explicitly or implicitly. Furthermore you can now access the main window and open Windows under the current application instance.
- What is the current behavior?
The current state calls exit when the main window was closed and there is no way to control this behavior
- What is the updated/expected behavior with this PR?
You are able to control when the application exits.
- How was the solution implemented (if it's not obvious)?
In the past all open windows where stored under Window and moved to Application with these changes. A ExitMode property was added to application and a SetExitMode method was added to the AppBuilder.

Checklist:

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

See  #1565

### CODE REVIEW

**Comment by Gillibald** [comment]
Yes it is the same. Just didn't know how to remove the merge commits. This will not happen again in the future. Now i know how to sync my fork properly.

File: src/Avalonia.Controls/Application.cs [hunk 1]
```diff
@@ -171,27 +209,95 @@ public virtual void Initialize()
...
+        public void Run(Window mainWindow)
+        {
+            if (_mainLoopCancellationTokenSource != null)
+            {
+                throw new Exception("Run should only called once");
+            }
+
+            _mainLoopCancellationTokenSource = new CancellationTokenSource();
+
+            Dispatcher.UIThread.InvokeAsync(
```
Comment (grokys) [inline src/Avalonia.Controls/Application.cs]: Why does this need to be done via an invoke?

File: src/Avalonia.Controls/ExitMode.cs [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// Copyright (c) The Avalonia Project. All rights reserved.
+// Licensed under the MIT license. See licence.md file in the project root for full license information.
+
+namespace Avalonia
+{
+    public enum ExitMode
```
Comment (grokys) [inline src/Avalonia.Controls/ExitMode.cs]: Could you add some XML documentation to this enum?

**Review by grokys — CHANGES_REQUESTED** [review]
~~Hmm, now if i set this in ControlCatalog.Desktop:~~

```
BuildAvaloniaApp().SetExitMode(ExitMode.OnMainWindowClose).Start<MainWindow>();
```

~~The application doesn't appear to start.~~

Edit: Hmm ignore that, it seems to be working now, not sure what was happening...

File: src/Avalonia.Controls/Application.cs (hunk 1 above)
Comment (Gillibald) [reply src/Avalonia.Controls/Application.cs]: I haven't found a better solution to show the main window and start up the main loop. The invoke queues the action for execution and gets executed when the main loop runs.

File: src/Avalonia.Controls/ExitMode.cs (hunk 2 above)
Comment (Gillibald) [reply src/Avalonia.Controls/ExitMode.cs]: Will add some comments

File: src/Avalonia.Controls/Application.cs (hunk 1 above)
Comment (grokys) [reply src/Avalonia.Controls/Application.cs]: I've tried changing this to:

```
            mainWindow.Show();
            MainWindow = mainWindow;
            Dispatcher.UIThread.MainLoop(_mainLoopCancellationTokenSource.Token);

```

And it _seems_ to work. What am I missing?

**Comment by Gillibald** [comment]
If everything works its fine. Not sure about side effects. My solution only runs after the main loop is running. Maybe show does the same. At some point something has to be invoked on the main thread. Is it possible to call run on a different thread or even Window.Show? My solution would avoid that issue.

File: src/Avalonia.Controls/Application.cs (hunk 1 above)
Comment (Gillibald) [reply src/Avalonia.Controls/Application.cs]: If everything works its fine. Not sure about side effects. My solution only runs after the main loop is running. Maybe show does the same. At some point something has to be invoked on the main thread. Is it possible to call run on a different thread or even Window.Show? My solution would avoid that issue.

Comment (Gillibald) [reply src/Avalonia.Controls/Application.cs]: Just tried what you did and it didn't work. My approach always runs. Don't know why yours work on your environment.

Comment (Gillibald) [reply src/Avalonia.Controls/Application.cs]: Have you tried it with Skia or Direct2D1? Yours works with Skia but not with Direct2D1 on my machine.

Comment (Gillibald) [reply src/Avalonia.Controls/Application.cs]: Dont understand it either has probably something to do with the. net runtime. With .Net Core I dont need invoke. Will remove the invoke call and just hope others don't have the problem.

**Review by grokys — APPROVED** [review]

**Comment by grokys** [comment]
Thanks @Gillibald !

**Pull request merged**